app/user.go

```diff
@@ -1375,6 +1375,8 @@ func (a *App) DeleteToken(token *model.Token) *model.AppError {
 }
 
 func (a *App) UpdateUserRoles(userId string, newRoles string, sendWebSocketEvent bool) (*model.User, *model.AppError) {
+	a.InvalidateCacheForUser(userId)
```

Comment:
IMO, this does not look to be the right fix. IIUC, our objective is to invalidate the cache after `userUpdate, err := a.Srv().Store.User().Update(user, true)` is called in `PatchBot`. But with this change, what we are doing here is invalidating the cache every time we call `GetUser`. I don't think `GetUser` has an inherent problem in getting the cached value by itself. Otherwise we would not have cached the user at all.

I believe the right solution is to invalidate the cache _after_ updating the user in `PatchBot`. That is the right order because this is "supposed" to behave like a write-through cache (though it's not), so all writes must go through the cache first, and then the store. And that is how it is done in the rest of the codebase. Technically, there is still a bug here because there is a period of time after updating and before invalidating, the cache will return stale data, but this is by design.

I have tested this patch locally with your PR and tested that it passes:
```diff
diff --git a/app/bot.go b/app/bot.go
index 09f96d326..9b53a4e5d 100644
--- a/app/bot.go
+++ b/app/bot.go
@@ -89,6 +89,7 @@ func (a *App) PatchBot(botUserId string, botPatch *model.BotPatch) (*model.Bot,
        if err != nil {
                return nil, err
        }
+       a.InvalidateCacheForUser(user.Id)
 
        ruser := userUpdate.New
        a.sendUpdatedUserEvent(*ruser)
diff --git a/app/user.go b/app/user.go
index 0c6971239..7fe7ae111 100644
--- a/app/user.go
+++ b/app/user.go
@@ -1375,7 +1375,7 @@ func (a *App) DeleteToken(token *model.Token) *model.AppError {
 }
 
 func (a *App) UpdateUserRoles(userId string, newRoles string, sendWebSocketEvent bool) (*model.User, *model.AppError) {
-       a.InvalidateCacheForUser(userId)
 
        user, err := a.GetUser(userId)
        if err != nil {

```

Thank you so much for spending your time to dig into this ! And sorry for the multiple rounds of review on this.

Reply:
Ok, that makes sense. Thanks again Agniva!